File: src/solutions/2025/04/c/main.c
#include "main.h"
#include "run.h"
#include "utils_num.h"
#include "utils_str.h"
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

struct position {
   int32_t x;
   int32_t y;
};

static uint8_t countneighbour(const int32_t x, const int32_t y,
                              const int32_t max, const char *restrict input,
                              const int32_t size) {
   uint8_t adjacent = 0;
   for (int32_t nX = -1; nX <= 1; nX++)
      for (int32_t nY = -1; nY <= 1; nY++) {
         int32_t index = (max * (y + nY)) + (x + nX) + (y + nY);
         if (index < 0 || index >= size)
            continue;
         adjacent += input[index] == '@';
      }
   return adjacent - 1;
}

static char *solve(const char *restrict input, const int32_t isTest,
                   const int32_t p2) {
   uint8_t grid[32766] = {};
   int32_t size = strlen(input);
   int32_t max = 0;
   for (int32_t i = 0; i < size; i++) {
      if (input[i] == '\n') {
         break;
      }
      max++;
   }

   struct position candidates[32766] = {};
   int32_t ptr = 0;

   for (int32_t y = 0; y < max; y++)
      for (int32_t x = 0; x < max; x++) {
         int32_t index = max * y + x + y;
         if (input[index] == '@') {
            grid[index] = countneighbour(x, y, max, input, size);
            if (grid[index] < 4) {
               candidates[ptr].x = x;
               candidates[ptr].y = y;
               ptr++;
            }
         }
      }

   int32_t total = 0;
   if (!p2) {
      return num_tostr(ptr);
   }

   while (ptr) {
      total++;
      struct position pos = candidates[--ptr];
      int32_t i = max * pos.y + pos.x + pos.y;
      grid[i] = 0;
      for (int32_t nX = -1; nX <= 1; nX++) {
         for (int32_t nY = -1; nY <= 1; nY++) {
            int32_t index = max * (pos.y + nY) + pos.x + nX + pos.y + nY;
            if (index < 0 || index >= size)
               continue;
            grid[index]--;
            if (grid[index] == 3) {
               candidates[ptr].x = pos.x + nX;
               candidates[ptr].y = pos.y + nY;
               ptr++;
            }
         }
      }
   }

   return num_tostr(total);
}

static char *part1(const char *restrict input, const int32_t isTest) {
   return solve(input, isTest, 0);
}

static char *part2(const char *restrict input, const int32_t isTest) {
   return solve(input, isTest, 1);
}

int main(int argc, char *argv[]) {
   solution_options_t options = getoptions();
   return run(argc, argv, part1, part2, options);
}

File: src/solutions/2025/04/rust/main.rs
extern crate aoc_core;
use std::{env, iter::FromIterator};

fn get_options() -> aoc_core::options::SolutionOptions {
   aoc_core::options::SolutionOptions {
      ..Default::default()
   }
}

fn count_neighbour(x: usize, y: usize, max: usize, input: &[u8]) -> u8 {
   let mut adjacent = 0;
   for n_x in -1..=1 {
      for n_y in -1..=1 {
         let p_x = x as isize + n_x;
         let p_y = y as isize + n_y;
         let index = max as isize * p_y + p_x + p_y;
         if index < 0 || index >= input.len() as isize {
            continue;
         }
         if input[index as usize] == b'@' {
            adjacent += 1;
         }
      }
   }
   adjacent - 1
}

fn update_neighbour(
   x: usize,
   y: usize,
   max: usize,
   grid: &mut [u8],
   candidates: &mut Vec<(usize, usize)>,
) {
   for n_x in -1..=1 {
      for n_y in -1..=1 {
         let p_x = x as isize + n_x;
         let p_y = y as isize + n_y;
         let index = max as isize * p_y + p_x + p_y;
         if index < 0 || index >= grid.len() as isize {
            continue;
         }
         grid[index as usize] -= 1;
         if grid[index as usize] == 3 {
            candidates.push((x + n_x as usize, y + n_y as usize));
         }
      }
   }
}

fn solve(input: &str, _is_test: bool, p2: bool) -> String {
   let mut grid = vec![0 as u8; input.len()];
   let input_bytes = input.as_bytes();
   let max = input_bytes.iter().position(|c| *c == b'\n').unwrap();

   let mut stack: Vec<(usize, usize)> = vec![];

   for y in 0..max {
      for x in 0..max {
         let index = max * y + x + y;
         if input_bytes[index] == b'@' {
            grid[index] = count_neighbour(x, y, max,input_bytes);
            if grid[index] < 4 {
               stack.push((x, y));
            }
         }
      }
   }

   if !p2 {
      return stack.len().to_string();
   }

   let mut total = 0;
   while stack.len() > 0 {
      let (x, y) = stack.pop().unwrap();
      grid[max * y + x + y] = 0;
      total += 1;
      update_neighbour(x, y, max, &mut grid, &mut stack);
   }

   total.to_string()
}

fn part_1(input: &str, is_test: bool) -> String {
   solve(input, is_test, false)
}

fn part_2(input: &str, is_test: bool) -> String {
   solve(input, is_test, true)
}

fn main() {
   let args = Vec::from_iter(env::args());
   aoc_core::run(
      args.iter().map(|x| x.as_ref()).collect::<Vec<_>>(),
      part_1,
      part_2,
      get_options(),
   );
}
